src/main/include/log4cxx/logger.h

```diff
@@ -617,13 +617,13 @@ class LOG4CXX_EXPORT Logger :
 
 		@throws RuntimeException if all levels are null in the hierarchy
 		*/
-		virtual const LevelPtr getEffectiveLevel() const;
+		virtual const LevelPtr& getEffectiveLevel() const;
 
 		/**
 		Return the the LoggerRepository where this
 		<code>Logger</code> is attached.
 		*/
-		log4cxx::spi::LoggerRepositoryWeakPtr getLoggerRepository() const;
+		log4cxx::spi::LoggerRepository* getLoggerRepository() const;
```

Comment:
I will measure the impact with 'm->priv->repository' a weak pointer. With  'm->priv->repository' a raw pointer, LOG4CXX_322 is still a probem.

However, with 'm->priv->repository' a weak pointer, getLoggerRepository() should return a spi::LoggerRepositoryPtr (i.e. getLoggerRepository() should call lock()), so the caller does not need to call lock().

Reply:
The per thread throughput is less half the single thread throughput when using a std::weak_ptr in Logger (Results are in #19)

Also, LOGCXX-322 is still a problem when using a std::weak_ptr in Logger.

Reply:
I agree that using the raw pointer would be faster, but what I'm saying is why don't we store the raw pointer and the weak pointer?  That lets the API stay stable, plus gives us the benefits of speed since we only use the raw pointer.

pseudo-code:
Logger.h:
```
private:
  log4cxx::spi::LoggerRepositoryWeakPtr repository;
  log4cxx::spi::LoggerRepository* repository_raw;
```
Logger.cpp:
```
void Logger::setHierarchy(spi::LoggerRepositoryWeakPtr repository1)
{
	this->repository = repository1;
        this->repository_raw = repository1.lock().get();
}

bool Logger::isTraceEnabled() const
{
	if (!repository_raw || repository_raw->isDisabled(Level::TRACE_INT))
	{
		return false;
	}
        return true;
}
```

Reply:
Given the std::weak_ptr does not fix LOGCXX-322, why is it is useful?

Reply:
Even if it doesn't completely fix it, there are at least two reasons that I can think of:
* Since it is part of the public API, it forces clients to use smart pointers(better expressing the intent) and keeps it similar to other parts of the API
* Keeps things stable for users who need to get the repository from the logger

Reply:
I expect that there's also a better way to fix this that requires some bigger architecture changes; I don't know what those are at the moment, but hopefully we can fix it for next_stable.

Reply:
OK, I agree  forceing clients to use smart pointers better expresses the intent.

So, I will change getLoggerRepository() to
```
log4cxx::spi::LoggerRepositoryPtr getLoggerRepository() const
```
OK?

Reply:
sounds good to me.